### PR TITLE
Adds "Add Prayer" Mutation to App

### PR DESCRIPTION
**Pull request by redreceipt**
## DESCRIPTION

Add ability to add prayers.

### How do I test this PR?

Boot her up, add some prayers!

---

> I am affirming this is my _best_ work ([Ecclesiastes 9:10](https://www.bible.com/bible/97/ECC.9.10.MSG))

## TODO

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [DEV-XXX](#url)
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android if applicable
- [ ] Set two relevant reviewers

## REVIEW

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

---

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js [hunk 1]
```diff
@@ -116,13 +122,15 @@ describe('PrayerRequest', () => {
   it('gets all public prayer requests by campus', async () => {
     const query = `
       query {
-        getPublicPrayerRequestsByCampus(campusId: 16) {
+        getPublicPrayerRequestsByCampus(campus: Campus:772fcb6087247ebad630814e2ce0cd16) {
```
Comment (conrad-vanl) [inline packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: So I noticed most of the graphql queries y'all have added follow this format of "getX". This is somewhat of an anti-pattern in graphql. Think of a query as a node in the graph - they represent an object, not a function. A more graphql-esque name for this would be something like `prayerRequests` or `publicPrayerRequests`

The "functional" naming comes in when you create a query with the `query` syntax:

```
query getPublicPrayerRequestsByCampus {
   prayerRequests(campus: "blah") {
      ...attr
   }
}
```

Since most of the queries you've added follow this `getX` naming structure, I wouldn't change it in this PR, but when y'all PR against core, we will be evaluating the schema structure, so food for thought.

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js [hunk 2]
```diff
@@ -337,13 +359,15 @@ describe('PrayerRequest', () => {
   it('creates a new prayer', async () => {
     const query = `
       mutation {
-        addPublicPrayerRequest(FirstName: "Test", LastName: "Bro", Text: "Jesus Rocks", CampusId: 16, CategoryId: 1, IsAnonymous: "True") {
+        addPublicPrayerRequest(FirstName: "Test", LastName: "Bro", Text: "Jesus Rocks", CampusId: Campus:772fcb6087247ebad630814e2ce0cd16, CategoryId: 1, IsAnonymous: true) {
```
Comment (conrad-vanl) [inline packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: when this is PR'd against core we'll need to convert these arguments from `PascalCase` to `camelCase`, and if `CategoryId` is a Rock ID, it needs to be changed to an Apollos ID.

**Review by conrad-vanl — COMMENTED** [review]
Code is looking pretty good. Left a few comments but nothing I think has to be addressed for this individual PR. I can't run/test on my own device locally, so this is strictly a code review.

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js (hunk 1 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: Yes love this! I was wondering how to make the client code simpler and I had this exact question "why doesn't my code line up semantically with Apollo's examples" this explains it.

@IsaacHardy I think we make this change, not before Thursday, but at some point to line up with Apollo and Apollos :)

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js (hunk 2 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: Yes on the camel case. Not sure what you mean about `CategoryId` but I do find it odd that it's the only variable we are hardcoding to 2 everywhere.

File: packages/apollos-data-prayer/src/prayer-requests/schema.js [hunk 3]
```diff
@@ -3,18 +3,18 @@ import gql from 'graphql-tag';
...
   extend type Mutation {
     addPublicPrayerRequest(
-      CampusId: Int!
+      CampusId: String!
       CategoryId: Int!
       Text: String!
       FirstName: String!
       LastName: String
-      IsAnonymous: String
+      IsAnonymous: Boolean
```
Comment (IsaacHardy) [inline packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: If I recall from my work on this, the `IsAnonymous` is a dumb property that stores 'True' and 'False' as strings and not booleans. What was the reason for reverting this to a boolean?

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 4]
```diff
@@ -1,7 +1,9 @@
 import React from 'react';
-import { Query } from 'react-apollo';
+import { Query, Mutation } from 'react-apollo';
 import { get } from 'lodash';
-import getUserProfile from 'newspringchurchapp/src/tabs/connect/getUserProfile';
+import GET_USER_PROFILE from 'newspringchurchapp/src/tabs/connect/getUserProfile';
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: for the record, we aren't using this type of capitalization anywhere else in the app for queries or mutation.

File: packages/apollos-data-prayer/src/prayer-requests/schema.js (hunk 3 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: because a bit is smaller than a five-byte character ;)

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 4 above)
Comment (redreceipt) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: Yea I saw that, but the convention makes sense and I think we should adopt it, not only because it's what Apollo prescribes, but so you don't confuse it with the function argument on the Mutation child render. Like this:

```
<Mutation mutation={getPrayers}>
    {(getPrayers) => (
        <Stuff />
    )}
</Mutation>
```

They are two completely different things but that's a little confusing as what's happening.

@conrad-vanl thoughts on why we avoided this convention all over the app?

File: packages/apollos-data-prayer/src/prayer-requests/schema.js (hunk 3 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: bools should *always* be used when they can be. It makes more sense to a developer trying to write client code too.

Comment (IsaacHardy) [reply packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: I don't disagree, I'm just telling you how this is handle on Rock's end which cannot be changed. My guess is that the IsAnonymous functionality is now broken.

Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: Nah dawg, I fixed that in the data source function.

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js (hunk 2 above)
Comment (conrad-vanl) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: the value `2` is probably directly tied to something in Rock. If so, we need to abstract it away from Rock into something that makes sense in the context of Apollos. If the `categoryId` is `2` everywhere, it should probably just be removed from the mutation.

Comment (conrad-vanl) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: we use `parseGlobalId` and such for `id`s elsewhere...that's what I mean.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 4 above)
Comment (conrad-vanl) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: 🤷‍♂

File: packages/apollos-data-prayer/src/prayer-requests/schema.js (hunk 3 above)
Comment (IsaacHardy) [reply packages/apollos-data-prayer/src/prayer-requests/schema.js:17]: Sweet.

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js (hunk 2 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js]: yep agreed

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js [hunk 5]
```diff
@@ -116,13 +121,16 @@ describe('PrayerRequest', () => {
   it('gets all public prayer requests by campus', async () => {
     const query = `
       query {
-        getPublicPrayerRequestsByCampus(campusId: 16) {
+        getPublicPrayerRequestsByCampus(campusId: "Campus:4f68015ba18662a7409d1219a4ce013e") {
```
Comment (richarddubay) [inline packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js:124]: This campus id returns `3` in your snapshot instead of `16` like it is everywhere else. Not sure that's a big deal because I know you're technically mocking that data. Just thought it looked funny.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 6]
```diff
@@ -11,12 +13,57 @@ class AddPrayerFormConnected extends React.Component {
...
   render() {
     return (
-      <Query query={getUserProfile} fetchPolicy={'cache-only'}>
-        {({ data }) => (
-          <AddPrayerForm
-            avatarSource={get(data, 'currentUser.profile.photo', { uri: null })}
-            {...this.props}
-          />
+      <Query query={GET_USER_PROFILE} fetchPolicy={'cache-only'}>
+        {(queryObj) => (
```
Comment (richarddubay) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: If all you're using here is `data`, why hide it behind `queryObj`?

File: packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js (hunk 5 above)
Comment (IsaacHardy) [reply packages/apollos-data-prayer/src/prayer-requests/__tests__/resolver.test.js:124]: He is specifically return 16 in his mock.

File: packages/apollos-data-prayer/src/prayer-requests/data-source.js [hunk 7]
```diff
@@ -102,17 +103,19 @@ export default class PrayerRequest extends RockApolloDataSource {
...
-        CampusId,
+        CampusId: parseInt(parseGlobalId(CampusId).id, 10),
         IsPublic: true,
         RequestedByPersonAliasId: primaryAliasId,
         IsActive: true,
         IsApproved: true,
         EnteredDateTime: new Date().toJSON(), // Required by Rock
       });
       // Sets the attribute value "IsAnonymous" on newly created prayer request
+      // TODO: we should combine this so network doesn't die and someone's prayer is left un-anonymous
```
Comment (IsaacHardy) [inline packages/apollos-data-prayer/src/prayer-requests/data-source.js:114]: What are you thinking here?

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js [hunk 8]
```diff
@@ -51,56 +51,59 @@ const SwitchContainer = styled({
...
+      onSubmit={onSubmit}
+    >
+      {({ handleChange, handleBlur, handleSubmit, values }) => (
+        <ModalView {...props}>
+          <FlexedSafeAreaView forceInset={{ top: 'always', bottom: 'never' }}>
+            <ShrinkingView behavior={'padding'}>
+              <HeaderView>
+                {/* TODO: margin will be fixed pending core #667 */}
+                <Avatar
+                  source={values.anonymous ? null : avatarSource}
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js:67]:
```suggestion
                  source={values.anonymous && values ? null : avatarSource}
```

This protects against value being undefined.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 4 above)
Comment (IsaacHardy) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: Lets change this to camel case for now. Just so it stays consistent.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 6 above)
Comment (IsaacHardy) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]:
```suggestion
        {({ data }) => (
```

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 9]
```diff
@@ -11,12 +13,57 @@ class AddPrayerFormConnected extends React.Component {
...
+              });
+            }}
+          >
+            {(addPrayer) => (
+              <AddPrayerForm
+                onSubmit={(values) => {
+                  addPrayer({
+                    variables: {
+                      campusID: get(
+                        queryObj.data,
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]:
```suggestion
                       data,
```

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 10]
```diff
@@ -11,12 +13,57 @@ class AddPrayerFormConnected extends React.Component {
...
+                        'currentUser.profile.campus.id'
+                      ),
+                      // TODO: make this dynamic
+                      categoryID: 2,
+                      text: values.prayer,
+                      firstName: get(
+                        queryObj.data,
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]:
```suggestion
                       data,
```

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 11]
```diff
@@ -11,12 +13,57 @@ class AddPrayerFormConnected extends React.Component {
...
+                        'currentUser.profile.firstName'
+                      ),
+                      lastName: get(
+                        queryObj.data,
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]:
```suggestion
                       data,
```

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js [hunk 12]
```diff
@@ -11,12 +13,57 @@ class AddPrayerFormConnected extends React.Component {
...
+                        'currentUser.profile.lastName'
+                      ),
+                      isAnonymous: values.anonymous,
+                    },
+                  });
+                  // TODO: wonder if this should be tied to the ModalView onClose method???
+                  this.props.navigation.pop();
+                }}
+                avatarSource={get(queryObj.data, 'currentUser.profile.photo', {
```
Comment (IsaacHardy) [inline packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]:
```suggestion
                avatarSource={get(data, 'currentUser.profile.photo', {
```

**Review by IsaacHardy — CHANGES_REQUESTED** [review]

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 6 above)
Comment (redreceipt) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: yea...I don't like it either. It's being used another scope below so I didn't know of a better way to do it.

File: packages/apollos-data-prayer/src/prayer-requests/data-source.js (hunk 7 above)
Comment (redreceipt) [reply packages/apollos-data-prayer/src/prayer-requests/data-source.js:114]: We should set the attribute to anonymous on the same post call we used to make the create the prayer. that way if there's a network failure in between calls, someone's prayer won't be left public.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js (hunk 8 above)
Comment (redreceipt) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js:67]: Don't think `values` would ever be `undefined` as they are coming from `Formik` so it'll always be something.

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js (hunk 9 above)
Comment (redreceipt) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerFormConnected.js]: I don't like it either but I can't do that because another argument is defined in the scope below as `data`

File: packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js (hunk 8 above)
Comment (IsaacHardy) [reply packages/newspringchurchapp/src/prayer/AddPrayer/AddPrayerForm/AddPrayerForm.js:67]: Cool.

**Review by IsaacHardy — APPROVED** [review]
